crates/bevy_pbr/src/render/light.rs

```diff
@@ -317,43 +333,71 @@ pub fn extract_lights(
     *previous_spot_lights_len = spot_lights_values.len();
     commands.insert_or_spawn_batch(spot_lights_values);
 
-    for (
-        entity,
-        directional_light,
-        visible_entities,
-        cascades,
-        cascade_config,
-        frusta,
-        transform,
-        view_visibility,
-        maybe_layers,
-        volumetric_light,
-    ) in &directional_lights
-    {
-        if !view_visibility.get() {
-            continue;
-        }
+    directional_lights.iter().for_each(
+        |(
+            entity,
+            directional_light,
+            visible_entities,
+            cascades,
+            cascade_config,
+            frusta,
+            transform,
+            view_visibility,
+            maybe_layers,
+            volumetric_light,
+        )| {
+            if !view_visibility.get() {
+                return;
+            }
 
-        // TODO: As above
-        let render_visible_entities = visible_entities.clone();
-        commands.get_or_spawn(entity).insert((
-            ExtractedDirectionalLight {
-                color: directional_light.color.into(),
-                illuminance: directional_light.illuminance,
-                transform: *transform,
-                volumetric: volumetric_light.is_some(),
-                shadows_enabled: directional_light.shadows_enabled,
-                shadow_depth_bias: directional_light.shadow_depth_bias,
-                // The factor of SQRT_2 is for the worst-case diagonal offset
-                shadow_normal_bias: directional_light.shadow_normal_bias * std::f32::consts::SQRT_2,
-                cascade_shadow_config: cascade_config.clone(),
-                cascades: cascades.cascades.clone(),
-                frusta: frusta.frusta.clone(),
-                render_layers: maybe_layers.unwrap_or_default().clone(),
-            },
-            render_visible_entities,
-        ));
-    }
+            // TODO: re-use the memory
```

Comment:
We could probably query for the existing entity hash maps and reuse them if they exist.

Reply:
yeah , thats the future plan ,if we got a retain render world .Most of the existing extraction logic can directly update components in place instead of reinserting them